### PR TITLE
[CARBONDATA-4019]Fix CDC merge failure join expression made of AND/OR expressions.

### PR DESCRIPTION
**Pull request by akashrn5**
### Why is this PR needed?
1. In CDC ,when the join expression contains AND/OR expression, then it failed with CAST exception
 2. when multiple columns are present in join condition, then not all columns are considered to check as 
bucket columns to repartition.

 ### What changes were proposed in this PR?
1. Instead of directly casting to equalTo expression to get join column, collect all attributes
2. check if all join columns present in bucket columns and then only repartition the data.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4260/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2515/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2553/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala [hunk 1]
```diff
@@ -106,18 +106,22 @@ case class CarbonMergeDataSetCommand(
     // decide join type based on match conditions
     val joinType = decideJoinType
 
-    val joinColumn = mergeMatches.joinExpr.expr.asInstanceOf[EqualTo].left
-      .asInstanceOf[UnresolvedAttribute].nameParts.tail.head
-    // repartition the srsDs, if the target has bucketing and the bucketing column and join column
-    // are same
+    val joinColumns = mergeMatches.joinExpr.expr.collect {
+      case unresolvedAttribute: UnresolvedAttribute if unresolvedAttribute.nameParts.nonEmpty =>
+        unresolvedAttribute.nameParts.tail.head
```
Comment (kunal642) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:111]: please write a comment explaining this logic

Comment (akashrn5) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/mutation/merge/CarbonMergeDataSetCommand.scala:117]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/4318/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/2568/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**